Fix flaky profiler and test_callback_simple RPC tests

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41287 Fix flaky profiler and test_callback_simple RPC tests**

Profiler tests that test profiling with builtin functions and `test_callback_simple` test has been broken for a while. This diff fixes that by preferring c10 ops to non-c10 ops in our operation matching logic.

The result of this is that these ops go through the c10 dispatch and thus have profiling enabled. For `test_callback_simple` this results in the effect that we choose `aten::add.Tensor` over `aten::add.Int` which fixes the type issue.

Differential Revision: [D22489197](https://our.internmc.facebook.com/intern/diff/D22489197/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22489197/)!

rohan-varma committed:
Fix flaky profiler and test_callback_simple RPC tests

Profiler tests that test profiling with builtin functions and `test_callback_simple` test has been broken for a while. This diff fixes that by preferring c10 ops to non-c10 ops in our operation matching logic.

The result of this is that these ops go through the c10 dispatch and thus have profiling enabled. For `test_callback_simple` this results in the effect that we choose `aten::add.Tensor` over `aten::add.Int` which fixes the type issue.

Differential Revision: [D22489197](https://our.internmc.facebook.com/intern/diff/D22489197/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22489197/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix flaky profiler and test_callback_simple RPC tests"


Profiler tests that test profiling with builtin functions and `test_callback_simple` test has been broken for a while. This diff fixes that by preferring c10 ops to non-c10 ops in our operation matching logic.

The result of this is that these ops go through the c10 dispatch and thus have profiling enabled. For `test_callback_simple` this results in the effect that we choose `aten::add.Tensor` over `aten::add.Int` which fixes the type issue.

Differential Revision: [D22489197](https://our.internmc.facebook.com/intern/diff/D22489197/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22489197/)!

[ghstack-poisoned]

lw reviewed (approved):
I'm not familiar with the code but the change is simple and looks safe as it only changes the order so to me it looks good! But maybe someone who knows the JIT dispatch better can provide more insight...

mrshenli commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -74,13 +77,20 @@ std::shared_ptr<Operator> matchBuiltinOp(
+      // torch.add) so a worst-case linear search should not incur significant
+      // extra overhead.
+      if (op->isC10Op()) {
+        return op;
```
Thanks for fixing! And thanks for digging into the bottom of this error!

This op-matching logic does need a thorough revamp.

rohan-varma committed:
Update on "Fix flaky profiler and test_callback_simple RPC tests"


Profiler tests that test profiling with builtin functions and `test_callback_simple` test has been broken for a while. This diff fixes that by preferring c10 ops to non-c10 ops in our operation matching logic.

The result of this is that these ops go through the c10 dispatch and thus have profiling enabled. For `test_callback_simple` this results in the effect that we choose `aten::add.Tensor` over `aten::add.Int` which fixes the type issue.

Differential Revision: [D22489197](https://our.internmc.facebook.com/intern/diff/D22489197/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22489197/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix flaky profiler and test_callback_simple RPC tests"


Profiler tests that test profiling with builtin functions and `test_callback_simple` test has been broken for a while. This diff fixes that by preferring c10 ops to non-c10 ops in our operation matching logic.

The result of this is that these ops go through the c10 dispatch and thus have profiling enabled. For `test_callback_simple` this results in the effect that we choose `aten::add.Tensor` over `aten::add.Int` which fixes the type issue.

Differential Revision: [D22489197](https://our.internmc.facebook.com/intern/diff/D22489197/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22489197/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix flaky profiler and test_callback_simple RPC tests"


Profiler tests that test profiling with builtin functions and `test_callback_simple` test has been broken for a while. This diff fixes that by preferring c10 ops to non-c10 ops in our operation matching logic.

The result of this is that these ops go through the c10 dispatch and thus have profiling enabled. For `test_callback_simple` this results in the effect that we choose `aten::add.Tensor` over `aten::add.Int` which fixes the type issue.

Differential Revision: [D22489197](https://our.internmc.facebook.com/intern/diff/D22489197/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22489197/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix flaky profiler and test_callback_simple RPC tests"


Profiler tests that test profiling with builtin functions and `test_callback_simple` test has been broken for a while. This diff fixes that by preferring c10 ops to non-c10 ops in our operation matching logic.

The result of this is that these ops go through the c10 dispatch and thus have profiling enabled. For `test_callback_simple` this results in the effect that we choose `aten::add.Tensor` over `aten::add.Int` which fixes the type issue.

Differential Revision: [D22489197](https://our.internmc.facebook.com/intern/diff/D22489197/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22489197/)!

[ghstack-poisoned]
